flink-connectors/flink-connector-datagen-test/pom.xml

```diff
@@ -0,0 +1,89 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements.  See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership.  The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License.  You may obtain a copy of the License at
+  ~
+  ~       http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+		 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+		 xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+	<modelVersion>4.0.0</modelVersion>
+
+	<parent>
+		<groupId>org.apache.flink</groupId>
+		<artifactId>flink-connectors</artifactId>
+		<version>1.18-SNAPSHOT</version>
+	</parent>
+
+	<artifactId>flink-connector-datagen-tests</artifactId>
```

Comment:
hmm. This isn't great is it. In a way it points to a deeper issue that I need module A to implement something but module B to test it.
I'm worried that this approach might no longer work in 2.0, depending on how the API module refactorings will go.

Right now I'm tending towards deprecating the existing methods in the `StreamExEnv` and just have connector-datagen have some factory methods for adding creating these sources/sinks.

Reply:
Definitely not great. I did not dig deeper, but on the surface it looks a bit strange that `flink-arcitecture-tests` require all of these test dependencies on major project modules https://github.com/apache/flink/blob/7d29e1beb721ac28fe4b803bce6e8af931e62044/flink-architecture-tests/pom.xml#L70 while mostly implementing a generic reflection-based framework. This seems to make it very prone to introduce cycles wherever it it used.

Reply:
Since `TestCodeArchitectureTestBase` uses `ITCaseRules` I assume you intent to move it too? This would pull `flink-test-utils`, `flink-connector-test-utils` and `junit-jupiter-api` with it into `architecture-tests-base`.

Reply:
junit would be fine in my mind, the others not so much :/
Could we rework the ITCaseRules to just check by canonical class name? :thinking:

Reply:
So basically rewrite methods  like `isAnnotatedWithExtendWithUsingExtension` and `areStaticFinalOfTypeWithAnnotation` that currently take `Class` (MiniClusterExtension.class)  as parameter to instead take a `String` ("MiniClusterExtension")?  I can give it a shot. Do you maybe know a trick how can I run all Arch tests at once? I'll need to make sure exactly the same classes match the predicate after the change.

Reply:
You could try `mvn verify -Dtest=TestCodeArchitectureTest -Dsurefire.failIfNoSpecifiedTests=false`

Reply:
>  that currently take Class (MiniClusterExtension.class) as parameter to instead take a String ("MiniClusterExtension")?

yup. But I _think_ it has to be the fully-qualified class name (org.apache.flink...MiniClusterExtension)

Reply:
I just realized that even after resolving the architect-tests cycle, we have the cycle simply because the `DataGeneratorSourceITCase` needs to use `StreamExecutionEnvironment`:
https://github.com/apache/flink/blob/81ed6c649e1a219c457628c54a7165d75b803474/flink-connectors/flink-connector-datagen/src/test/java/org/apache/flink/connector/datagen/source/DataGeneratorSourceITCase.java#L81
:/

Reply:
I believe the only options we have are:
1.  Drop the `DataGen ITCase`                                                      - not good
2. Move the `DataGen ITCase` into `flink-streaming-java`        - weird
3. Continue with the `flink-connector-datagen-tests` split       - not great, not terrible (c)
4. Move `#fromCollection()` methods into `flink-datagen`       - not great because it breaks the familiar pattern, but since a lot of things will probably change in 2.0 anyway, this should be fine. An open question is whether we want to also add dependency from `flink-datagen` to `flink-streaming-java`. This will improve usability, otherwise it will have look somewhat like this: 
```
Integer data = ...;
DataGenSource<Integer> collectionSource = DataGenSource.fromCollection(data);
DataStreamSource<Integer> source = env.fromSource(collectionSource,
                        WatermarkStrategy.forMonotonousTimestamps(),
                        "Collection source")
                .forceNonParallel();
```
Especially the necessity for the `forceNonParallel()/setParallelism(1)` call is concerning because it is easy to forget.

Reply:
I think I'd prefer either option 2 or 3.

1 is just not an option isn't it.
4 is just a nosedive in usability, and we've had enough of that recently.

2/3 are similar in that the datagen ITCase is no longer next to the datagen connector, which seems unavoidable.
Either would be fine with me I think. Maybe a slight tendency toward the additional module to to at least keep the modules closer.

Reply:
Perfect, thanks for you input. The "not great, not terrible (c)" option it is then, I also had a slight preference towards it. I'll add comments to the pom to make it clear why this module exists.

Reply:
Small question: I already prepared the rework of `flink-architecture-tests` towards using the fully-qualified names instead of these somewhat redundant dependencies for class objects. If we go for option (3) we do not immediately need it here. Would you like to see this refactoring integrated in order to minimize the dependencies surface nonetheless? If so, I'll open a separate ticket and a PR.

Reply:
> Would you like to see this refactoring integrated in order to minimize the dependencies surface regardless of that? If so, I'll open a separate ticket and a PR.

Should be a separate ticket :+1:

Reply:
Resolved https://issues.apache.org/jira/browse/FLINK-33199